File: Classes/HMAC.m
//
//  HMAC.m
//  Pods
//
//  Created by Ryan Maxwell on 29/06/17.
//
//

#import "HMAC.h"
#import <CommonCrypto/CommonCrypto.h>

@implementation HMAC

+ (NSString *)signData:(NSData *)data withKey:(NSString *)key usingAlgorithm:(HMACAlgorithm)algorithm {
    const char *hKey = [key cStringUsingEncoding:NSUTF8StringEncoding];
    const void *hData = data.bytes;
    unsigned char cHMAC[[HMAC digestLengthForAlgorithm:algorithm]];
    
    CCHmacAlgorithm ccAlgorithm = [HMAC CCHmacAlgorithmForAlgorithm: algorithm];
    CCHmac(ccAlgorithm, hKey, strlen(hKey), hData, data.length, cHMAC);
    NSData *resultData = [[NSData alloc] initWithBytes:cHMAC
                                                length:sizeof(cHMAC)];
    char *utf8;
    utf8 = (char *)[resultData bytes];
    NSMutableString *hex = [NSMutableString string];
    for (int i = 0; i < resultData.length; i++) {
        [hex appendFormat:@"%02X" , *(utf8 + i) & 0x00FF];
    }
    return [hex copy];
}

+ (CCHmacAlgorithm)CCHmacAlgorithmForAlgorithm:(HMACAlgorithm)algorithm {
    switch (algorithm) {
        case HMACAlgorithmMD5:
            return kCCHmacAlgMD5;
        case HMACAlgorithmSHA1:
            return kCCHmacAlgSHA1;
        case HMACAlgorithmSHA224:
            return kCCHmacAlgSHA224;
        case HMACAlgorithmSHA256:
            return kCCHmacAlgSHA256;
        case HMACAlgorithmSHA384:
            return kCCHmacAlgSHA384;
        case HMACAlgorithmSHA512:
            return kCCHmacAlgSHA512;
    }
}

+ (int)digestLengthForAlgorithm:(HMACAlgorithm)algorithm {
    switch (algorithm) {
        case HMACAlgorithmMD5:
            return CC_MD5_DIGEST_LENGTH;
        case HMACAlgorithmSHA1:
            return CC_SHA1_DIGEST_LENGTH;
        case HMACAlgorithmSHA224:
            return CC_SHA224_DIGEST_LENGTH;
        case HMACAlgorithmSHA256:
            return CC_SHA256_DIGEST_LENGTH;
        case HMACAlgorithmSHA384:
            return CC_SHA384_DIGEST_LENGTH;
        case HMACAlgorithmSHA512:
            return CC_SHA512_DIGEST_LENGTH;
    }
}

@end

File: Classes/HMAC.swift
//
//  HMAC.swift
//  LetterboxdAPI
//
//  Created by Ryan Maxwell on 20/06/17.
//  Copyright © 2017 Letterboxd. All rights reserved.
//

import Foundation

public struct HMAC {

	public enum Algorithm {
		case sha1
		case md5
		case sha256
		case sha384
		case sha512
		case sha224

		public var algorithm: CCHmacAlgorithm {
			switch self {
			case .md5: return CCHmacAlgorithm(kCCHmacAlgMD5)
			case .sha1: return CCHmacAlgorithm(kCCHmacAlgSHA1)
			case .sha224: return CCHmacAlgorithm(kCCHmacAlgSHA224)
			case .sha256: return CCHmacAlgorithm(kCCHmacAlgSHA256)
			case .sha384: return CCHmacAlgorithm(kCCHmacAlgSHA384)
			case .sha512: return CCHmacAlgorithm(kCCHmacAlgSHA512)
			}
		}

		public var digestLength: Int {
			switch self {
			case .md5: return Int(CC_MD5_DIGEST_LENGTH)
			case .sha1: return Int(CC_SHA1_DIGEST_LENGTH)
			case .sha224: return Int(CC_SHA224_DIGEST_LENGTH)
			case .sha256: return Int(CC_SHA256_DIGEST_LENGTH)
			case .sha384: return Int(CC_SHA384_DIGEST_LENGTH)
			case .sha512: return Int(CC_SHA512_DIGEST_LENGTH)
			}
		}
	}

	// MARK: - Signing

	public static func sign(data: Data, withKey: Data, algorithm: Algorithm) -> String {
		let signature = UnsafeMutablePointer<UInt8>.allocate(capacity: algorithm.digestLength)
		defer { signature.deallocate(capacity: algorithm.digestLength) }

		data.withUnsafeBytes { dataBytes in
			key.withUnsafeBytes { keyBytes in
				CCHmac(algorithm.algorithm, keyBytes, key.count, dataBytes, data.count, signature)
			}
		}

		var hash = ""
		for i in 0 ..< algorithm.digestLength {
			hash += String(format: "%02X", signature[i])
		}
		return hash
	}
}
